Add `LocalForm` to make testing with form data more ergonomic

ELD opened the pull request:
This adds the testing improvement described in #1591.

The transformation of the form fields into the resulting body data is a bit clumsy and could probably use some improvement/cleanup. I believe the multipart form translation is correct, but there's only one, narrow test case that confirms it works as expected.

Additionally, I was unable to get something like this to work without ugly ergonomics:
```rust
let client = Client::tracked(rocket).unwrap();
let response = client.post("/")
    .form(&[("field", "value"), ("is it", "a cat?")])
    .dispatch();
```

I might be missing something, but attempting to do something like this:
```rust
impl<'v, F: Into<ValueField<'v>>, I: Iterator<Item = F>> From<I> for LocalForm { /* .. */ }
```

requires the caller to call the method in a way similar to this:

```rust
client.post("/")
    .form(&[("field", "value"), ("is it", "a cat?")].iter().cloned());
```

due to a slice reference not being coerced into an iterator and when it is iterated over, it returns `&(&str, &str)`. `ValueField` doesn't have a present `From<..>` impl for `&(&str, &str)`, so this is not possible.

Also, docs are likely not complete/elaborate enough at this stage, but an attempt at documenting relevant methods/structs was made in this first cut.

Please pick it apart so I can make the implementation better! :)

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+    /// The full content-type for this form.
+    pub fn content_type(&self) -> ContentType {
+        if self.contains_data_field() {
+            return "multipart/form-data; boundary=X-BOUNDARY"
```
use `ContentType::with_params()`.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    fn contains_data_field(&self) -> bool {
+        self
+            .0
+            .iter()
+            .any(|field| matches!(field, LocalField::Data(..)))
```
```suggestion
        self.0.iter().any(|field| matches!(field, LocalField::Data(..)))
```

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    fn contains_data_field(&self) -> bool {
```
I don't think this needs to be a standalone function.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    /// A percent-encoded `name` and `value`.
+    pub fn raw_field(mut self, name: &'v RawStr, value: &'v RawStr) -> Self {
```
Take an `Into<Cow<'_, RawStr>>`.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
```
I didn't have it in my original API, but this field name needs to be a parameter to the function. We can't just set it to `file`.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
```
Same here.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    fn format_simple(&self) -> Vec<u8> {
+        self.0.iter().fold(Vec::new(), |mut acc, field| {
+            match field {
+                LocalField::Value(name, value) => {
+                    acc.push(format!("{}={}", Uri::percent_encode(&format!("{}", name)), Uri::percent_encode(value)));
+                    return acc
+                },
+                _ => acc,
+            }
+        }).join("&").as_bytes().to_vec()
```
Ideally we allocate a total of 1 `Vec<u8>` here, pass a pointer it around, and use `write!` to fill it up.

Here's an example of how to do that:

```rust
use std::io;

fn f(buf: &mut dyn io::Write) -> io::Result<()> {
    write!(buf, "hello, {}", "there")?;
    Ok(())
}

pub fn main() {
    let mut buf = vec![];
    let _ = f(&mut buf);
}
```

I'd like the this code to have _zero_ calls to `format!`, or `to_string()`, or similar.

SergioBenitez commented on core/lib/src/local/form.rs:
```diff
@@ -0,0 +1,222 @@
+                    acc.push(format!("Content-Disposition: form-data; name=\"{}\"; filename=\"{}\"", name, file_name.unwrap_or("")));
+                    acc.push(format!("Content-Type: {}", content_type));
+                    acc.push("".to_string());
+                    acc.push(format!("{}", String::from_utf8_lossy(data)));
```
This is wrong; `data` may not be UTF-8, and this will incorrectly format the data. This shouldn't be a `String`, but part of the byte buffer.

SergioBenitez reviewed (changes requested):
The code is handling percent-encoding incorrectly. It's decoding when it shouldn't be and failing to encode when it should be. Here are the conditions:

  * If a provided string _is_ percent-encoded, it should be left alone.
  * If a provided string _is not_ percent-encoded, it should be percent-encoded.

Specifically, note that for some `s`, `encode(decode(s)) != s` (for example, `%`).

Note that we have the `UriDisplay<Query>` trait that already knows how to do this for a bunch of types. We should probably use it here. Perhaps better type signatures are:

```rust
pub fn field<N, V>(mut self, name: N, value: V) -> Self
    where N: Into<NameBuf<'v>>, V: UriDisplay<Query>,
```

Let's start by fixing the incorrect code and style and we'll go from there.

ELD commented:
@SergioBenitez This should be ready for another pass.

I updated the style to better match Rocket's. I think it's mostly there, but there might be some minor style issues remaining.

I changed the writing of the body data to a single `Vec<u8>` with the `write!` macro. There's a couple places where I call `format!`. There's probably a better way than what I did, so any help there would be great!

Error handling is a little weird as it currently silently outputs an empty buffer if the writing failed for some reason. This probably isn't a good user experience because an error would result in empty body data for the form request. Any pointers on how to do better error handling would be great!

slyons commented:
Has anyone had a chance to look at this yet?
